src/refdb_reftable.c

```diff
@@ -0,0 +1,513 @@
+	reftable_writer_arg arg = {
+		refs, logs, nrefs, nlogs, stack_next_update_index(backend->stack)
+	};
+
+	if (stack_add(backend->stack, reftable_writer, &arg) < 0)
+		return -1;
+
+	return 0;
+}
+
+static int reftable_exists(
+	int *exists,
+	git_refdb_backend *_backend,
+	const char *ref_name)
+{
+	reftable_backend *backend = GIT_CONTAINER_OF(_backend, reftable_backend, parent);
+	struct ref_record ref = { 0 };
+	int error;
+
+	if ((error = stack_read_ref(backend->stack, ref_name, &ref)) < 0)
+		return error;
+	*exists = (error == 0);
+
+	ref_record_clear(&ref);
+	return 0;
+}
+
+static int reftable_lookup(
+	git_reference **out,
+	git_refdb_backend *_backend,
+	const char *ref_name)
+{
+	reftable_backend *backend = GIT_CONTAINER_OF(_backend, reftable_backend, parent);
+	struct ref_record ref;
+	int error;
+
+	if ((error = stack_read_ref(backend->stack, ref_name, &ref)) < 0)
+		return error;
+	if (error > 0)
+		return GIT_ENOTFOUND;
+
+	return reftable_reference_from_record(out, &ref);
+}
+
+static int reftable_iterator_next(git_reference **out, git_reference_iterator *_it)
+{
+	reftable_iterator *it = GIT_CONTAINER_OF(_it, reftable_iterator, parent);
+	struct ref_record ref;
+	int error;
+
+	while (1) {
+		if ((error = iterator_next_ref(it->iter, &ref)) < 0)
+			return error;
+		if (error > 0)
+			return GIT_ITEROVER;
+		if (it->glob && wildmatch(it->glob, ref.ref_name, 0) != 0)
+			continue;
+		return reftable_reference_from_record(out, &ref);
+	}
+
+	return GIT_ITEROVER;
+}
+
+static int reftable_iterator_next_name(const char **out, git_reference_iterator *_it)
+{
+	reftable_iterator *it = GIT_CONTAINER_OF(_it, reftable_iterator, parent);
+	struct ref_record ref;
+	int error;
+
+	while (1) {
+		if ((error = iterator_next_ref(it->iter, &ref)) < 0)
+			return error;
+		if (error > 0)
+			return GIT_ITEROVER;
+		if (it->glob && wildmatch(it->glob, ref.ref_name, 0) != 0)
+			continue;
+
+		*out = ref.ref_name;
+		return 0;
+	}
+
+	return GIT_ITEROVER;
+}
+
+static void reftable_iterator_free(git_reference_iterator *_it)
+{
+	reftable_iterator *it = GIT_CONTAINER_OF(_it, reftable_iterator, parent);
+	merged_table_free(it->table);
+	iterator_destroy(&it->iter);
+	git__free(it);
+}
+
+static int reftable_iterator_new(
+	git_reference_iterator **out, git_refdb_backend *_backend, const char *glob)
+{
+	reftable_backend *backend = GIT_CONTAINER_OF(_backend, reftable_backend, parent);
+	struct merged_table *table = NULL;
+	reftable_iterator *it = NULL;
+	int error;
+
+	table = stack_merged_table(backend->stack);
+	GIT_ERROR_CHECK_ALLOC(table);
+
+	it = git__calloc(1, sizeof(*it));
+	GIT_ERROR_CHECK_ALLOC(it);
+	it->parent.next = reftable_iterator_next;
+	it->parent.next_name = reftable_iterator_next_name;
+	it->parent.free = reftable_iterator_free;
+	it->table = table;
+	it->glob = git__strdup(glob);
+	GIT_ERROR_CHECK_ALLOC(it->glob);
+
+	if ((error = merged_table_seek_ref(table, &it->iter, "")) < 0)
```

Comment:
It was a bit hard to discover how to actually get the first entry of a ref table. Documenting this in the different `seek_ref` and `seek_log` calls would probably help.

Reply:
good point, I'll get to that.